collector/collector.go

```diff
@@ -0,0 +1,875 @@
+						copy(rns, ns)
+
+						rns[indexOfDynamicElement+lengthOfNsPrefix].Value = size
+						metric := plugin.Metric{
+							Timestamp: time.Now(),
+							Namespace: rns,
+							Data:      utils.GetValueByNamespace(hugetlb, []string{mt.Namespace.Strings()[len(ns)-1]}),
+							Tags:      mt.Tags,
+							Config:    mt.Config,
+							Version:   PLUGIN_VERSION,
+						}
+
+						metrics = append(metrics, metric)
+					}
+				} else {
+					metric := plugin.Metric{
+						Timestamp: time.Now(),
+						Namespace: ns,
+						Data:      utils.GetValueByNamespace(c.containers[rid].Stats.Cgroups.HugetlbStats, mt.Namespace.Strings()[len(ns)-2:]),
+						Tags:      mt.Tags,
+						Config:    mt.Config,
+						Version:   PLUGIN_VERSION,
+					}
+					metrics = append(metrics, metric)
+				}
+
+			case "io_merged_recursive":
+				numOfDevices := len(c.containers[rid].Stats.Cgroups.BlkioStats.IoMergedRecursive) - 1
+				if metricName[0] == "*" {
+					for deviceID, imr := range c.containers[rid].Stats.Cgroups.BlkioStats.IoMergedRecursive {
+						rns := make([]plugin.NamespaceElement, len(ns))
+						copy(rns, ns)
+
+						rns[indexOfDynamicElement+lengthOfNsPrefix].Value = strconv.Itoa(deviceID)
+						metric := plugin.Metric{
+							Timestamp: time.Now(),
+							Namespace: rns,
+							Data:      utils.GetValueByNamespace(imr, mt.Namespace.Strings()[len(ns)-1:]),
+							Tags:      mt.Tags,
+							Config:    mt.Config,
+							Version:   PLUGIN_VERSION,
+						}
+						metrics = append(metrics, metric)
+					}
+				} else {
+					deviceID, err := strconv.Atoi(metricName[0])
+					if err != nil {
+						return nil, fmt.Errorf("In metric %s the given device id is invalid, err=%v", strings.Join(mt.Namespace.Strings(), "/"), err)
+					}
+					if deviceID > numOfDevices || deviceID < 0 {
+						return nil, fmt.Errorf("In metric %s the given device id is invalid, expected value in range 0-%d", strings.Join(mt.Namespace.Strings(), "/"), numOfDevices)
+					}
+
+					metric := plugin.Metric{
+						Timestamp: time.Now(),
+						Namespace: ns,
+						Data:      utils.GetValueByNamespace(c.containers[rid].Stats.Cgroups.BlkioStats.IoMergedRecursive[deviceID], mt.Namespace.Strings()[len(ns)-1:]),
+						Tags:      mt.Tags,
+						Config:    mt.Config,
+						Version:   PLUGIN_VERSION,
+					}
+					metrics = append(metrics, metric)
+				}
+			case "io_service_bytes_recursive":
+				numOfDevices := len(c.containers[rid].Stats.Cgroups.BlkioStats.IoServiceBytesRecursive) - 1
+				if metricName[0] == "*" {
+					for deviceID, isbr := range c.containers[rid].Stats.Cgroups.BlkioStats.IoServiceBytesRecursive {
+						rns := make([]plugin.NamespaceElement, len(ns))
+						copy(rns, ns)
+
+						rns[indexOfDynamicElement+lengthOfNsPrefix].Value = strconv.Itoa(deviceID)
+						metric := plugin.Metric{
+							Timestamp: time.Now(),
+							Namespace: rns,
+							Data:      utils.GetValueByNamespace(isbr, mt.Namespace.Strings()[len(ns)-1:]),
```

Comment:
The same as above, probably taking[len(ns)-1] instead [len(ns)-1:] will be more appropriate.